### PR TITLE
Implement Permutation with CK

### PR DESCRIPTION
**Pull request by CongMa13**

### CODE REVIEW

File: samples/common.hpp [hunk 1]
```diff
@@ -35,10 +35,10 @@
...
     }
 
-#define HIPTENSOR_FREE_HOST(ptr) \
-    if(ptr != nullptr)           \
-    {                            \
-        free(ptr);               \
+#define HIPTENSOR_FREE_HOST(ptr)           \
+    if(ptr != nullptr)                     \
+    {                                      \
+        CHECK_HIP_ERROR(hipHostFree(ptr)); \
```
Comment (cgmillette) [inline samples/common.hpp:41]: Good catch

File: test/utils.hpp [hunk 2]
```diff
@@ -191,6 +209,19 @@ std::pair<bool, double> compareEqual(DDataType const* deviceD,
     return std::make_pair(retval, max_relative_error);
 }
 
+template <typename DDataType>
+double getEpsilon()
+{
+    if(std::is_same_v<DDataType, _Float16>)
+    {
+        return 0.0009765625; // numeric_limits<_Float16>::epsilon() => 0
```
Comment (cgmillette) [inline test/utils.hpp]: Should we include the std::numeric_limits<_Float16> extension from rocWMMA?

File: library/src/permutation/permutation_ck_col_impl.hpp
Comment (cgmillette) [inline library/src/permutation/permutation_ck_col_impl.hpp:1]: Is this file specific to column major implementation? Or can we generalize?

**Review by cgmillette — COMMENTED** [review]
We might have to think about how to treat the __half datatype. 
At some level we will have to wrap __half as _Float16 as they follow the same IEEE format.

**Comment by cgmillette** [comment]
@CongMa13 we should coordinate with CI about updating the CK package so that pre-checkin should pass

File: library/src/permutation/permutation_ck_col_impl.hpp
Comment (CongMa13) [reply library/src/permutation/permutation_ck_col_impl.hpp:1]: Updated the PR and hiptensor supports both row and col major layout. 

Users can use CMake option `HIPTENSOR_DATA_LAYOUT_COL_MAJOR` to choose the layout.

File: test/utils.hpp (hunk 2 above)
Comment (CongMa13) [reply test/utils.hpp]: yeah, definitely. I will copy type related files from wmma to hiptensor.

**Review by cgmillette — DISMISSED** [review]
LGTM!

File: test/utils.hpp (hunk 2 above)
Comment (CongMa13) [reply test/utils.hpp]: In commit b76153
- Copied xfloat32, type_traits and types_ext from wmma to hiptensor
- replace this hard code with type_traits of f16

File: library/src/include/config.hpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+#if !defined(HIPTENSOR_ARCH_GFX940)
+#define HIPTENSOR_ARCH_GFX940 0
+#endif
+#if !defined(HIPTENSOR_ARCH_GFX941)
+#define HIPTENSOR_ARCH_GFX941 0
+#endif
+#if !defined(HIPTENSOR_ARCH_GFX942)
+#define HIPTENSOR_ARCH_GFX942 0
+#endif
+#if !defined(HIPTENSOR_ARCH_GFX1100)
```
Comment (cgmillette) [inline library/src/include/config.hpp]: No current support for gfx11

File: library/src/include/config.hpp [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+#define HIPTENSOR_ARCH_GFX1101 0
+#endif
+#if !defined(HIPTENSOR_ARCH_GFX1102)
+#define HIPTENSOR_ARCH_GFX1102 0
+#endif
+#if !defined(HIPTENSOR_ARCH_HOST)
+#define HIPTENSOR_ARCH_HOST 0
+#endif
+
+///
```
Comment (cgmillette) [inline library/src/include/config.hpp]: from here

File: library/src/include/config.hpp [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+#endif
+#if !defined(HIPTENSOR_WAVE32_MODE)
+#define HIPTENSOR_WAVE32_MODE 0
+#endif
+#if !defined(HIPTENSOR_BLOCK_DIM_16_SUPPORTED)
+#define HIPTENSOR_BLOCK_DIM_16_SUPPORTED 0
+#endif
+#if !defined(HIPTENSOR_BLOCK_DIM_32_SUPPORTED)
+#define HIPTENSOR_BLOCK_DIM_32_SUPPORTED 0
+#endif
```
Comment (cgmillette) [inline library/src/include/config.hpp]: to here: can remove

File: library/src/include/config.hpp [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+#if HIPTENSOR_NO_HALF || (!HIPTENSOR_NO_HALF && defined(__HIP_NO_HALF_CONVERSIONS__))
+#define HIPTENSOR_TESTS_NO_HALF 1
+#else
+#define HIPTENSOR_TESTS_NO_HALF 0
+#endif // !HIPTENSOR_NO_HALF && defined(__HIP_NO_HALF_CONVERSIONS__)
+
+///
+/// Sanity checks
+///
+#if HIPTENSOR_ARCH_GFX11
```
Comment (cgmillette) [inline library/src/include/config.hpp]: can remove

File: library/src/include/config.hpp [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    static_assert((bool)(HIPTENSOR_WAVE32_MODE) && !(bool)(HIPTENSOR_WAVE64_MODE),
+                  "hipTensor supports only wave32 for gfx11 arch");
+    static_assert((bool)(HIPTENSOR_BLOCK_DIM_16_SUPPORTED)
+                      && !(bool)(HIPTENSOR_BLOCK_DIM_32_SUPPORTED),
+                  "hipTensor supports only block size of 16 for gfx11 arch");
+#endif
+
+#if HIPTENSOR_ARCH_GFX9
```
Comment (cgmillette) [inline library/src/include/config.hpp]: can remove

File: library/src/include/native_types.hpp [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef HIPTENSOR_NATIVE_TYPES_HPP
+#define HIPTENSOR_NATIVE_TYPES_HPP
+
+#if !defined(__HIPCC_RTC__)
```
Comment (cgmillette) [inline library/src/include/native_types.hpp]: Can remove support for hipRTC

File: library/src/include/native_types.hpp [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+    using int8_t   = ::int8_t;
+    using uint8_t  = ::uint8_t;
+    using int16_t  = ::int16_t;
+    using uint16_t = ::uint16_t;
+    using int32_t  = ::int32_t;
+    using uint32_t = ::uint32_t;
+    using int64_t  = ::int64_t;
+    using uint64_t = ::uint64_t;
+    using index_t  = ::int32_t;
```
Comment (cgmillette) [inline library/src/include/native_types.hpp]: should use index_t as int64_t

File: library/src/include/native_types.hpp [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+#else
+
+    using int8_t   = __hip_internal::int8_t;
```
Comment (cgmillette) [inline library/src/include/native_types.hpp]: can remove

File: library/src/include/type_traits.hpp [hunk 11]
```diff
@@ -0,0 +1,749 @@
...
+#ifndef HIPTENSOR_TYPE_TRAITS_HPP
+#define HIPTENSOR_TYPE_TRAITS_HPP
+
+#if !defined(__HIPCC_RTC__)
+
+#include <cfloat>
+
+#else
+
+#define FLT_EPSILON __FLT_EPSILON__
```
Comment (cgmillette) [inline library/src/include/type_traits.hpp]: can remove

File: library/src/include/type_traits.hpp [hunk 12]
```diff
@@ -0,0 +1,749 @@
...
+
+    } // namespace detail
+} // namespace hiptensor
+
+///////////////////////////////////////////////////////////
+/////////////  std replacements for hipRTC  ///////////////
+///////////////////////////////////////////////////////////
+#if defined(__HIPCC_RTC__)
+namespace std
+{
```
Comment (cgmillette) [inline library/src/include/type_traits.hpp]: Can remove

File: library/src/include/type_traits.hpp [hunk 13]
```diff
@@ -0,0 +1,749 @@
...
+    template <typename T>
+    using decay_t = typename decay<T>::type;
+
+} // namespace std
+#endif
+
+namespace std
+{
+#if defined(__HIPCC_RTC__)
+    using uint16_t = hiptensor::uint16_t;
```
Comment (cgmillette) [inline library/src/include/type_traits.hpp]: can remove

**Review by cgmillette — APPROVED** [review]

**Pull request merged**